storage/grpc_client.go

```diff
@@ -1512,10 +1545,78 @@ func (r *gRPCReader) Read(p []byte) (int, error) {
 		r.leftovers = content[n:]
 	}
 	r.seen += int64(n)
+	r.updateCRC(p[:n])
 
 	return n, nil
 }
 
+// WriteTo writes all the data requested by the Reader into w, implementing
+// io.WriterTo.
+func (r *gRPCReader) WriteTo(w io.Writer) (int64, error) {
+	// The entire object has been read by this reader, check the checksum if
+	// necessary and return EOF.
+	if r.size == r.seen || r.zeroRange {
+		if err := r.runCRCCheck(); err != nil {
+			return 0, err
+		}
+		return 0, io.EOF
```

Comment:
I noticed this above in the read too, but I wonder if these should be storage specific errors so users have more context? Or maybe wrapping happens at a higher layer. `fmt.Errof("storage: something something: %w", io.EOF)`. Then your checks would need to be changed to `errors.Is(err, io.EOF)`, but I think that change should happen regardless.

Reply:
Yeah I think there is still more we can do to have more coherently wrapped errors. it's been on the back burner.